### PR TITLE
Handle history, implement header Last-Event-Id

### PR DESCRIPTION
**Pull request by bpaquet**
Implement Last-Event-Id.
Keep events for between `HISTORY_MIN_DURATION` and `HISTORY_MIN_DURATION` * 2.

Data are stored in two ETS table, which are written alternatively.

### CODE REVIEW

File: neurow/lib/neurow/receiver_shard_manager.ex [hunk 1]
```diff
@@ -0,0 +1,76 @@
+defmodule Neurow.ReceiverShardManager do
```
Comment (achouippe) [inline neurow/lib/neurow/receiver_shard_manager.ex:1]: Maybe it should be a [DynamicSupervisor](https://hexdocs.pm/elixir/DynamicSupervisor.html) ? So ReceiverShard will be children of the ReceiverShardManager in the supervision tree?

File: neurow/lib/neurow/receiver_shard.ex [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        user_topic,
+        {:pubsub_message, message_id, message}
+      )
+
+    true = :ets.insert(table_1, {String.to_atom(user_topic), {message_id, message}})
+    {:noreply, {table_0, table_1}}
+  end
+
+  @impl true
+  def handle_call({:rotate}, _, {table_0, table_1}) do
```
Comment (achouippe) [inline neurow/lib/neurow/receiver_shard.ex:56]: From what I it, that seems to be a good practice to implement a public method directly on the GenServer to expose the rotate mechanism, so client do not have to directly submit messages through the GenServer API.

```
def rotate(pid) do
  GenServer.call(pid, {:rotate})
end 
```

File: neurow/lib/neurow/receiver_shard_manager.ex [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    Enum.map(0..(@shards - 1), fn shard ->
+      fun.({shard, Neurow.ReceiverShard.build_name(shard)})
+    end)
+  end
+
+  def rotate do
+    Stats.inc_history_rotate()
+
+    all_pids(fn {_, pid} ->
+      GenServer.call(pid, {:rotate})
```
Comment (achouippe) [inline neurow/lib/neurow/receiver_shard_manager.ex:44]: GenServer.call is synchronous. If we do not need any answer from the targeted genservers, maybe using async calls with handle_cast/GenServer.cast will speedup the process a bit ?

File: neurow/lib/neurow/receiver_shard.ex [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    {:ok, {table_0, table_1}}
+  end
+
+  # Read from the current process, not from GenServer process
+  def get_history(shard, topic) do
+    table_0 = table_name(shard, 0)
+    table_1 = table_name(shard, 1)
+
+    result =
+      :ets.lookup(table_0, String.to_atom(topic)) ++
```
[the remark this answers is not in the document]
Comment (bpaquet) [reply neurow/lib/neurow/receiver_shard.ex:36]: We do not need atom :) I'm not sure why I added a conversion here.

File: neurow/lib/neurow/receiver_shard.ex (hunk 2 above)
Comment (bpaquet) [reply neurow/lib/neurow/receiver_shard.ex:56]: Done

File: neurow/lib/neurow/receiver_shard_manager.ex (hunk 3 above)
Comment (bpaquet) [reply neurow/lib/neurow/receiver_shard_manager.ex:44]: Done

**Review by achouippe — APPROVED** [review]
LGTM

**Pull request merged**